crates/astria-sequencer/src/checked_actions/ics20_withdrawal.rs

```diff
@@ -0,0 +1,702 @@
+use std::fmt::{
+    self,
+    Debug,
+    Formatter,
+};
+
+use astria_core::{
+    primitive::v1::{
+        asset::{
+            Denom,
+            IbcPrefixed,
+        },
+        Address,
+        Bech32,
+        ADDRESS_LEN,
+    },
+    protocol::{
+        memos::v1::Ics20WithdrawalFromRollup,
+        transaction::v1::action::Ics20Withdrawal,
+    },
+};
+use astria_eyre::{
+    anyhow_to_eyre,
+    eyre::{
+        bail,
+        ensure,
+        OptionExt as _,
+        Result,
+        WrapErr as _,
+    },
+};
+use cnidarium::{
+    StateRead,
+    StateWrite,
+};
+use ibc_types::core::channel::{
+    ChannelId,
+    PortId,
+};
+use penumbra_ibc::component::packet::{
+    IBCPacket,
+    SendPacketRead as _,
+    SendPacketWrite as _,
+    Unchecked,
+};
+use penumbra_proto::core::component::ibc::v1::FungibleTokenPacketData;
+use tracing::{
+    instrument,
+    Level,
+};
+
+use super::{
+    AssetTransfer,
+    TransactionSignerAddressBytes,
+};
+use crate::{
+    accounts::StateWriteExt as _,
+    address::StateReadExt as _,
+    app::StateReadExt as _,
+    bridge::{
+        StateReadExt as _,
+        StateWriteExt as _,
+    },
+    ibc::{
+        StateReadExt as _,
+        StateWriteExt as _,
+    },
+};
+
+pub(crate) struct CheckedIcs20Withdrawal {
+    action: Ics20Withdrawal,
+    withdrawal_address: [u8; ADDRESS_LEN],
+    bridge_address_and_rollup_withdrawal: Option<(Address, Ics20WithdrawalFromRollup)>,
+    ibc_packet: IBCPacket<Unchecked>,
+    tx_signer: TransactionSignerAddressBytes,
+}
+
+impl CheckedIcs20Withdrawal {
+    #[instrument(skip_all, err(level = Level::DEBUG))]
+    pub(super) async fn new<S: StateRead>(
+        action: Ics20Withdrawal,
+        tx_signer: [u8; ADDRESS_LEN],
+        state: S,
+    ) -> Result<Self> {
+        state
+            .ensure_base_prefix(&action.return_address)
+            .await
+            .wrap_err("failed to verify that return address has permitted base prefix")?;
+
+        ensure!(action.timeout_time != 0, "timeout time must be non-zero");
+        ensure!(action.amount > 0, "amount must be greater than zero");
+        let withdrawal_address = action
+            .bridge_address
+            .as_ref()
+            .map_or(tx_signer, |address| address.bytes());
+        let bridge_address_and_rollup_withdrawal = if let Some(bridge_address) =
+            action.bridge_address
+        {
+            state
+                .ensure_base_prefix(&bridge_address)
+                .await
+                .wrap_err("bridge address has an unsupported prefix")?;
+            let parsed_bridge_memo: Ics20WithdrawalFromRollup = serde_json::from_str(&action.memo)
+                .wrap_err("failed to parse memo for ICS bound bridge withdrawal")?;
```

Comment:
```suggestion
                .wrap_err("failed to parse memo for outgoing IBC bridge withdrawal")?;
```

Reply:
Done in [73ec272](https://github.com/astriaorg/astria/pull/2142/commits/73ec27210915e8a170768ff722a20d45297b4f81).